src/main/java/com/amazon/ion/impl/macro/Expression.kt

```diff
@@ -52,7 +52,21 @@ sealed interface Expression {
      * These expressions are the only ones that may be the output from the macro evaluator.
      * All [DataModelExpression]s are also valid to use as [TemplateBodyExpression]s and [EExpressionBodyExpression]s.
      */
-    sealed interface DataModelExpression : Expression, EExpressionBodyExpression, TemplateBodyExpression
+    sealed interface DataModelExpression : Expression, EExpressionBodyExpression, TemplateBodyExpression, ExpansionOutputExpression
+
+    /** Output of a macro expansion (internal to the macro evaluator) */
+    sealed interface ExpansionOutputExpressionOrContinue
+    /** Output of the macro evaluator */
+    sealed interface ExpansionOutputExpression : ExpansionOutputExpressionOrContinue
+
+    /**
+     * Indicates to the macro evaluator that the current expansion did not produce a value this time, but it may
+     * produce more expressions. The macro evaluator should request another expression from that macro.
+     */
+    data object ContinueExpansion : ExpansionOutputExpressionOrContinue
+
+    /** Signals the end of an expansion in the macro evaluator. */
+    data object EndOfExpansion : ExpansionOutputExpression
```

Comment:
🗺️ It was getting confusing to keep track of what `null` meant when returned from some of the internal functions of the macro evaluator. I added the singleton types `ContinueExpansion` and `EndOfExpansion` (and marker interfaces that combine them with `DataModelExpression`) so that we can be explicit about the result of a call to the macro expansion function instead of returning `null`.